File: gcp-microservices-demo/src/recommendationservice/recommendationservice/handler.py
import json
from flask import request, Response
import rest
import http
from otel import tracer
from opentelemetry.propagate import extract, inject
from opentelemetry.trace import SpanKind

def main():
    with tracer.start_as_current_span("handle request", context=extract(request.headers), kind=SpanKind.SERVER) as span:
        if request.method == "GET":   #ListRecommendations
            span.add_event("invoke ListRecommendations")
            try:
                user_id = request.args.get("user_id")
                product_ids = []
                raw_product_ids = request.args.get("product_ids")
                if raw_product_ids != None:
                    product_ids = raw_product_ids.split(",")
                req = rest.ListRecommendationsRequest(user_id, product_ids)
                headers = {}
                inject(headers)
                resp = rest.recommendationservice.listRecommendations(headers, req)
                resp_body = json.dumps(resp.toDict())
                span.add_event("successfully handle request")
                return Response(response=resp_body, status=http.HTTPStatus.OK)
            except Exception as e:
                print(e)
                span.add_event("an error occurred in ListRecommendations")
                return Response(status=http.HTTPStatus.BAD_REQUEST)
        else:
            print("methods other than GET are not supported")
            span.add_event("methods other than GET are not supported")
            return Response(status=http.HTTPStatus.BAD_REQUEST)

File: gcp-microservices-demo/src/currencyservice/currencyservice/handler.js
var rest = require('./rest/rest.js');
const pino = require('pino');

const logger = pino({
    name: 'currencyservice',
    messageKey: 'message',
    levelKey: 'severity',
    useLevelLabels: true
});

const api = require('@opentelemetry/api')
const { defaultTextMapGetter, ROOT_CONTEXT, SpanKind } = require('@opentelemetry/api');
const tracer = require('./tracer')('currencyservice');

var currencyservice = new rest.CurrencyService();

module.exports = async function(context) {
    const parentCtx = api.propagation.extract(ROOT_CONTEXT, context.request.headers, defaultTextMapGetter)
    const span = tracer.startSpan(
        'handle request',
        {
            kind: SpanKind.SERVER,
        },
        parentCtx,
    )

    if (context.request.method == "GET") {  //GetSupportedCurrencies
        try {
            span.addEvent("invoke GetSupportedCurrencies");
            var resp = currencyservice.getSupportedCurrencies();
            span.addEvent("successfully handle request");
            span.end();
            return {
                status: 200,
                body: resp
            }
        } catch(err) {
            logger.error(err);
            span.addEvent("an error occurred in GetSupportedCurrencies");
            span.end();
            return {
                status: 400,
                body: new rest.GetSupportedCurrenciesResponse()
            }
        }
    } else if (context.request.method == "POST") {    //Convert
        try {
            span.addEvent("invoke Convert");
            var req = new rest.CurrencyConversionRequest(
                new rest.Money(
                    context.request.body.from.currency_code,
                    context.request.body.from.units,
                    context.request.body.from.nanos
                ),
                context.request.body.to_code
            );
            var resp = currencyservice.convert(req);
            span.addEvent("successfully handle request");
            span.end();
            return {
                status: 200,
                body: resp
            }
        } catch(err) {
            logger.error(err);
            span.addEvent("an error occurred in Convert");
            span.end();
            return {
                status: 400,
                body: new rest.Money()
            }
        }
    } else {
        logger.error("methods other than GET and POST are not supported");
        span.addEvent("methods other than GET and POST are not supported");
        span.end();
        return {
            status: 400
        }
    }
}